src/server/transaction.cc

```diff
@@ -913,43 +847,57 @@ void Transaction::Execute(RunnableType cb, bool conclude) {
                                   : (coordinator_state_ & ~COORD_CONCLUDING);
   }
 
-  ExecuteAsync();
+  if ((coordinator_state_ & COORD_SCHED) == 0) {
+    ScheduleInternal();
+  }
 
+  DispatchHop();
   run_barrier_.Wait();
-
   cb_ptr_ = nullptr;
+
+  if (coordinator_state_ & COORD_CONCLUDING)
+    coordinator_state_ &= ~COORD_SCHED;
 }
 
 // Runs in coordinator thread.
-void Transaction::ExecuteAsync() {
-  DVLOG(1) << "ExecuteAsync " << DebugId();
+void Transaction::DispatchHop() {
+  DVLOG(1) << "DispatchHop " << DebugId();
   DCHECK_GT(unique_shard_cnt_, 0u);
   DCHECK_GT(use_count_.load(memory_order_relaxed), 0u);
   DCHECK(!IsAtomicMulti() || multi_->lock_mode.has_value());
-  DCHECK_LE(shard_data_.size(), 1024u);
+  DCHECK_LE(shard_data_.size(), 256u);
 
   // Hops can start executing immediately after being armed, so we
   // initialize the run barrier before arming, as well as copy indices
   // of active shards to avoid reading concurrently accessed shard data.
-  std::bitset<1024> poll_flags(0);
-  run_barrier_.Start(unique_shard_cnt_);
+  std::bitset<256> poll_flags(0);
```

Comment:
why 256?
I am asking because it's dangerously close to today hardware limits (see m6a.48xlarge)

Reply:
I wanted to make sure it has no effects on benchmarks (overprotection) 😆 But it, it's just 16 bytes...